src/System.IO.Ports/src/System/IO/Ports/SerialStream.Unix.cs

```diff
@@ -563,27 +624,231 @@ internal SafeFileHandle OpenPort(string portName)
+
+            return 0;
+        }
+
+        private unsafe int ProcessWrite(SerialStreamAsyncResult r)
+        {
+            ReadOnlySpan<byte> buff = r.Buffer.Span;
+            fixed (byte* bufPtr = buff)
+            {
+                // assumes dequeue-ing happens on a single thread
+                int numBytes = Interop.Sys.Write(_handle, bufPtr, buff.Length);
+
+                Interop.Error lastError = numBytes < 0 ? Interop.Sys.GetLastError() : 0;
+                if (numBytes > 0)
+                {
+                    r.ProcessBytes(numBytes);
+
+                    if (r.Buffer.Length == 0)
                     {
+                        r.Complete();
                     }
-                    base.Dispose(disposing);
+
+                    return numBytes;
+                }
+                // ignore EWOULDBLOCK since we handle timeout elsewhere
+                // numBytes == 0 means that there might be an error
+                else if (lastError != Interop.Error.SUCCESS && lastError != Interop.Error.EWOULDBLOCK)
+                {
+                    r.Complete(new IOException());
+                }
+            }
+
+            return 0;
+        }
+
+        // returns number of bytes read/written
+        private static int DoIORequest(ConcurrentQueue<SerialStreamAsyncResult> q, RequestProcessor op)
+        {
+            // assumes dequeue-ing happens on a single thread
+            while (q.TryPeek(out SerialStreamAsyncResult r))
+            {
+                if (r.IsCompleted)
+                {
+                    Debug.Assert(q.TryDequeue(out _));
+                    // take another item since we haven't processed anything
+                    continue;
+                }
+
+                int ret = op(r);
+                Debug.Assert(ret >= 0);
+
+                if (r.IsCompleted)
+                {
+                    Debug.Assert(q.TryDequeue(out _));
+                }
+
+                return ret;
+            }
+
+            return 0;
+        }
+
+        private unsafe void IOLoop()
+        {
+            bool eofReceived = false;
+            //bool readyForReceivedEvent = true;
+            long totalBytesRead = 0;
+            // last value of totalBytesRead + BytesToRead
+            long lastTotalBytesAvailable = BytesToRead;
+            while (IsOpen && !eofReceived && !_ioLoopFinished)
+            {
+                // PollEvents will wait if both flags are false
+                Interop.Sys.PollEvents events = PollEvents(1,
+                                                           pollReadEvents: _readQueue.Count > 0,
+                                                           pollWriteEvents: _writeQueue.Count > 0);
+
+                if (events.HasFlag(Interop.Sys.PollEvents.POLLNVAL) ||
+                    events.HasFlag(Interop.Sys.PollEvents.POLLERR))
+                {
+                    // bad descriptor or some other error we can't handle
+                    break;
```

Comment:
What happens to pending reads/writes in this case... they just hang?

Reply:
I slightly assumed that this happens only when fd gone bad - in that case we will immediately call Dispose which will complete those tasks. It probably makes sense to complete them in here though